### PR TITLE
Multiple features (NOTICE, LIST, logs)

### PR DESCRIPTION
**Pull request by delthas**
_Several commits in one PR for easier rebasing. See individual commits for details._

_Not that Github may show some commits in the wrong order (ordered by date rather than by parent)._

### CODE REVIEW

File: upstream.go [hunk 1]
```diff
@@ -860,3 +1128,14 @@ func (uc *upstreamConn) readMessages(ch chan<- upstreamIncomingMessage) error {
 func (uc *upstreamConn) SendMessage(msg *irc.Message) {
 	uc.outgoing <- msg
 }
+
+func (uc *upstreamConn) SendMessageLabeled(dc *downstreamConn, msg *irc.Message) {
+	if uc.labelsSupported {
+		if msg.Tags == nil {
+			msg.Tags = make(map[string]irc.TagValue)
+		}
+		msg.Tags["label"] = irc.TagValue(fmt.Sprintf("sd-%d-%d", dc.id, uc.nextLabelId))
```
Comment (delthas) [inline upstream.go]: Should this be obfuscated somehow? (I think not). Also, are uint64 counters OK for downstream ids and label ids? Either that or randomly generated UUIDs could work.

File: upstream.go [hunk 2]
```diff
@@ -697,11 +958,14 @@ func (uc *upstreamConn) handleMessage(msg *irc.Message) error {
 		}
 
 		uc.forEachDownstream(func(dc *downstreamConn) {
+			if downstreamId != 0 && downstreamId != dc.id {
+				return
+			}
```
Comment (delthas) [inline upstream.go]: This piece of code used to select a specific downstream id is used multiple times throughout the code, should we create a function for that? Maybe a new `forEachDownstream` that also takes an optional downstream id.

File: irc.go [hunk 3]
```diff
@@ -54,6 +54,89 @@ func (ms *modeSet) Apply(s string) error {
...
+	'l': modeTypeC,
+	'i': modeTypeD,
+	'm': modeTypeD,
+	'n': modeTypeD,
+	's': modeTypeD,
+	't': modeTypeD,
+}
+
+func ApplyChannelModes(modeTypes map[byte]channelModeType, modes map[byte]string, modeStr string, arguments ...string) error {
+	nextArgument := 0
```
Comment (emersion) [inline irc.go]: We should use a consistent naming: either argument or param. Here we have `arguments`, `param`, `nextArgument`.

File: irc.go [hunk 4]
```diff
@@ -54,6 +54,89 @@ func (ms *modeSet) Apply(s string) error {
...
+		}
+		if plusMinus != '+' && plusMinus != '-' {
+			return fmt.Errorf("malformed modestring %q: missing plus/minus", modeStr)
+		}
+
+		mt, ok := modeTypes[mode]
+		if !ok {
+			continue
+		}
+		if mt == modeTypeB || (mt == modeTypeC && plusMinus == '+') {
```
Comment (emersion) [inline irc.go]: This can probably be simplified to something among those lines:

```go
if plusMinus == '+' {
	var param string
	switch mt {
	case modeTypeB, modeTypeC:
		// some sentitive arguments […]
		if nextArgument < len(arguments) {
			…
		}
	case modeTypeC, modeTypeD:
		// Mode has no param
	}
	modes[mode] = param
} else {
	delete(modes, mode)
}
```

File: irc.go [hunk 5]
```diff
@@ -54,6 +54,89 @@ func (ms *modeSet) Apply(s string) error {
 	return nil
 }
 
+type channelModeType byte
+
+const (
+	modeTypeA channelModeType = iota
```
Comment (emersion) [inline irc.go]: Would be nice to have doc comments about what these are

File: irc.go [hunk 6]
```diff
@@ -54,6 +54,89 @@ func (ms *modeSet) Apply(s string) error {
...
+	'k': modeTypeB,
+	'l': modeTypeC,
+	'i': modeTypeD,
+	'm': modeTypeD,
+	'n': modeTypeD,
+	's': modeTypeD,
+	't': modeTypeD,
+}
+
+func ApplyChannelModes(modeTypes map[byte]channelModeType, modes map[byte]string, modeStr string, arguments ...string) error {
```
Comment (emersion) [inline irc.go]: Just like we have `userModes`, maybe it makes sense to have `type channelModes map[byte]string` with methods like `Apply` and `String`?

File: irc.go (hunk 3 above)
Comment (delthas) [reply irc.go]: Right. I've picked argument, for "mode argument" (as written in the specs)

File: irc.go (hunk 4 above)
Comment (delthas) [reply irc.go]: Not sure I can simply further. See `nextArgument++`.

File: irc.go (hunk 5 above)
Comment (delthas) [reply irc.go]: Added.

File: irc.go (hunk 6 above)
Comment (delthas) [reply irc.go]: OK. I'm not sure about "`String()`" though b/c we don't return a string but also a string array.

Comment (emersion) [reply irc.go]: What about `Format` or `Marshal`?

Comment (delthas) [reply irc.go]: `Format` is good, i'll do that

File: upstream.go [hunk 7]
```diff
@@ -508,6 +548,88 @@ func (uc *upstreamConn) handleMessage(msg *irc.Message) error {
...
+		if err != nil {
+			return err
+		}
+
+		firstMode := ch.modes == nil
+		ch.modes = make(map[byte]string)
+		if err := ch.modes.Apply(uc.availableChannelModes, modeStr, msg.Params[3:]...); err != nil {
+			return err
+		}
+		if firstMode {
```
Comment (emersion) [inline upstream.go]: We should probably send `RPL_CHANNELMODEIS` to downstream connections unconditionally. Otherwise clients may become out-of-sync.

File: upstream.go [hunk 8]
```diff
@@ -508,6 +548,88 @@ func (uc *upstreamConn) handleMessage(msg *irc.Message) error {
...
+		}
+		modeStr := ""
+		if len(msg.Params) > 1 {
+			modeStr = msg.Params[1]
+		}
+
+		uc.modes = ""
+		if err := uc.modes.Apply(modeStr); err != nil {
+			return err
+		}
```
Comment (emersion) [inline upstream.go]: Can we add a TODO: send RPL_UMODEIS to downstream connections when applicable?

File: downstream.go [hunk 9]
```diff
@@ -913,6 +879,49 @@ func (dc *downstreamConn) handleMessageRegistered(msg *irc.Message) error {
 				})
 			}
```
Comment (emersion) [inline downstream.go]: A `break` is missing here, to prevent the `!uc.isChannel(upstreamName)` check to send an error.

File: upstream.go [hunk 10]
```diff
@@ -346,11 +325,67 @@ func (uc *upstreamConn) handleMessage(msg *irc.Message) error {
...
+				negate = true
+				token = token[1:]
+			} else {
+				if i := strings.IndexByte(token, '='); i >= 0 {
+					parameter = token[:i]
+					value = token[i+1:]
+				}
+			}
+			if !negate {
+				switch parameter {
```
Comment (emersion) [inline upstream.go]: Calling `strings.ToUpper` would allow more liberal parsing.

File: downstream.go [hunk 11]
```diff
@@ -891,45 +891,75 @@ func (dc *downstreamConn) handleMessageRegistered(msg *irc.Message) error {
...
-			})
 		} else {
-			ch, ok := uc.channels[upstreamName]
-			if !ok {
-				return ircError{&irc.Message{
-					Command: irc.ERR_NOSUCHCHANNEL,
-					Params:  []string{dc.nick, name, "No such channel"},
-				}}
-			}
+			if modeStr != "" {
```
Comment (emersion) [inline downstream.go]: This change is unrelated to this commit. I dropped it while applying the patch to master.

File: downstream.go [hunk 12]
```diff
@@ -891,45 +891,75 @@ func (dc *downstreamConn) handleMessageRegistered(msg *irc.Message) error {
...
+		}
+
+		channels := strings.Split(msg.Params[0], ",")
+		for _, channel := range channels {
+			uc, upstreamChannel, err := dc.unmarshalEntity(channel)
+			if err != nil {
+				return err
+			}
+
+			ch, err := uc.getChannel(upstreamChannel)
```
Comment (emersion) [inline downstream.go]: I prefer to use `ch, ok := uc.channels[upstreamChannel]` when we want to handle both cases (the case where the channel exists, and the case where it doesn't). `getChannel` is just a helper to unify "channel not found" errors.

Comment (emersion) [reply downstream.go]: (TBH we could replace `getChannel` with an `errNoSuchChannel` variable)

File: upstream.go [hunk 13]
```diff
@@ -757,7 +757,7 @@ func (uc *upstreamConn) handleMessage(msg *irc.Message) error {
 			dc.SendMessage(&irc.Message{
 				Prefix:  dc.srv.prefix(),
 				Command: irc.RPL_ENDOFWHO,
-				Params:  []string{dc.nick, name, "End of WHO list"},
+				Params:  []string{dc.nick, name, "End of /WHO list"},
```
Comment (emersion) [inline upstream.go]: Unrelated change, moving it to an earlier commit while cherry-picking.

File: upstream.go [hunk 14]
```diff
@@ -984,6 +1002,8 @@ func (uc *upstreamConn) handleCapAck(name string, ok bool) error {
 			Command: "AUTHENTICATE",
 			Params:  []string{auth.Mechanism},
 		})
+	case "message-tags":
+		uc.tagsSupported = true
```
Comment (emersion) [inline upstream.go]: `= ok`

File: upstream.go [hunk 15]
```diff
@@ -888,6 +904,8 @@ func (uc *upstreamConn) handleMessage(msg *irc.Message) error {
 				Params:  []string{dc.marshalNick(uc, nick), dc.marshalChannel(uc, channel)},
 			})
 		})
+	case "TAGMSG":
+		// Ignore
```
Comment (emersion) [inline upstream.go]: The spec says this should be handled just like a `PRIVMSG` (ie. relayed). We should add a `// TODO: relay to downstream accepting message-tags`.

**Comment by emersion** [comment]
Cherry-picked all commits up to (and including) "Add downstream self WHO and WHOIS support".

To rebase:

```sh
git fetch upstream
git rebase -i upstream/master
# pick → drop all commits cherry-picked in master
```

File: upstream.go (hunk 1 above)
Comment (emersion) [reply upstream.go]: Yeah, I think all of this is fine.

File: upstream.go [hunk 16]
```diff
@@ -1044,6 +1074,8 @@ func (uc *upstreamConn) handleCapAck(name string, ok bool) error {
 		})
 	case "message-tags":
 		uc.tagsSupported = true
+	case "labeled-response":
+		uc.labelsSupported = true
```
Comment (emersion) [inline upstream.go]: `= ok`

File: upstream.go (hunk 2 above)
Comment (emersion) [reply upstream.go]: +1

File: upstream.go (hunk 15 above)
Comment (hhirtz) [reply upstream.go]: Here, you can apply this suggestion ;)

```suggestion
		// TODO: relay to downstream accepting message-tags
```

Comment (emersion) [reply upstream.go]: Please don't use GitHub suggestions. Those add fixup commits instead of amending the original commit.

Comment (hhirtz) [reply upstream.go]: Oh ok, didn't know.  I see how this is not wanted here.

**Comment by emersion** [comment]
Pushed all commits except the last one. Thanks!

File: upstream.go [hunk 17]
```diff
@@ -206,9 +206,28 @@ func (uc *upstreamConn) handleMessage(msg *irc.Message) error {
 	case "NOTICE":
 		uc.logger.Print(msg)
 
-		uc.forEachDownstream(func(dc *downstreamConn) {
-			dc.SendMessage(msg)
-		})
+		if msg.Prefix.User == "" && msg.Prefix.Host == "" {
+			uc.SendMessage(msg)
```
Comment (emersion) [inline upstream.go]: This bounces the message back to the upstream server. We probably want to send it to downstream instead, or am I missing something?

File: upstream.go [hunk 18]
```diff
@@ -206,9 +206,28 @@ func (uc *upstreamConn) handleMessage(msg *irc.Message) error {
...
+			var nick, text string
+			if err := parseMessageParams(msg, &nick, &text); err != nil {
+				return err
+			}
+
+			target := nick
+			if nick == uc.nick {
+				target = msg.Prefix.Name
+			}
+			uc.Logfn(target, "<%s> %s", msg.Prefix.Name, text)
```
Comment (emersion) [inline upstream.go]: This belongs to a later commit.

Comment (emersion) [reply upstream.go]: Ah, it seems like this has been removed in 3d8819ad8054f25aa83a83a23208f5a70c0b619f.

File: downstream.go [hunk 19]
```diff
@@ -1236,6 +1236,23 @@ func (dc *downstreamConn) handleMessageRegistered(msg *irc.Message) error {
...
+		}
+
+		for _, name := range strings.Split(targetsStr, ",") {
+			uc, upstreamName, err := dc.unmarshalEntity(name)
+			if err != nil {
+				return err
+			}
+
+			uc.SendMessage(&irc.Message{
+				Command: "PRIVMSG",
```
Comment (emersion) [inline downstream.go]: Shouldn't this be `NOTICE`?

File: user.go [hunk 20]
```diff
@@ -87,6 +87,15 @@ type user struct {
 	networks        []*network
 	nextUpstreamID  uint64 // use with atomic
 	downstreamConns []*downstreamConn
+
+	pendingListsLock sync.Mutex
```
Comment (emersion) [inline user.go]: Note to self: all messages (upstream and downstream) are processed in a per-user goroutine (I should write down the goroutine architecture). This means that the only reason why we need this lock is to clear the pending lists when an upstream connection is closed.

File: upstream.go [hunk 21]
```diff
@@ -1089,6 +1211,8 @@ func (uc *upstreamConn) handleMessage(msg *irc.Message) error {
 		// Ignore
 	case irc.RPL_MOTDSTART, irc.RPL_MOTD, irc.RPL_ENDOFMOTD:
 		// Ignore
+	case irc.RPL_LISTSTART:
```
Comment (emersion) [inline upstream.go]: Do we need to send an initial `RPL_LISTSTART` to clients?

Comment (delthas) [reply upstream.go]: Probably not, it's optional and marked as deprecated in RFC 2812

File: upstream.go (hunk 17 above)
Comment (delthas) [reply upstream.go]: Right, thats a mistake

File: downstream.go (hunk 19 above)
Comment (delthas) [reply downstream.go]: Right!

File: upstream.go [hunk 22]
```diff
@@ -204,9 +204,30 @@ func (uc *upstreamConn) handleMessage(msg *irc.Message) error {
 	case "NOTICE":
 		uc.logger.Print(msg)
 
-		uc.forEachDownstream(func(dc *downstreamConn) {
-			dc.SendMessage(msg)
-		})
+		if msg.Prefix.User == "" && msg.Prefix.Host == "" { // server message
```
Comment (emersion) [inline upstream.go]: Can the IRC server send a NOTICE without any prefix? This is what Freenode sends for a server NOTICE:

```
:verne.freenode.net NOTICE * :*** Looking up your hostname...
```

File: upstream.go [hunk 23]
```diff
@@ -204,9 +204,30 @@ func (uc *upstreamConn) handleMessage(msg *irc.Message) error {
...
+				dc.SendMessage(msg)
+			})
+		} else { // regular user NOTICE
+			var nick, text string
+			if err := parseMessageParams(msg, &nick, &text); err != nil {
+				return err
+			}
+
+			target := nick
+			if nick == uc.nick {
```
Comment (emersion) [inline upstream.go]: Why do we have this special case? If the target is ourselves, why should we change the target to the source of the NOTICE message?

Comment (emersion) [reply upstream.go]: Oh, this is only used for logs. Can we move this whole code block in the commit adding support for logs? (`git rebase -i`, edit both commits, cut the block from this commit, `git rebase --continue`, paste the block to the next one)

File: config/config.go [hunk 24]
```diff
@@ -19,18 +20,25 @@ type Server struct {
...
 	SQLSource string
+	LogPath   string
 }
 
 func Defaults() *Server {
 	hostname, err := os.Hostname()
 	if err != nil {
 		hostname = "localhost"
 	}
+	logPath, err := os.UserHomeDir()
```
Comment (emersion) [inline config/config.go]: Not sure about this default. soju is designed to be multi-user, so logging in ~ sounds weird. I'd prefer having logs disabled by default and enabled in the config file.

File: upstream.go [hunk 25]
```diff
@@ -1344,3 +1395,47 @@ func (uc *upstreamConn) SendMessageLabeled(downstreamID uint64, msg *irc.Message
 	}
 	uc.SendMessage(msg)
 }
+
+// TODO: handle moving logs when a network name changes, when support for this is added
+func (uc *upstreamConn) Logfn(entity string, format string, a ...interface{}) {
```
Comment (emersion) [inline upstream.go]: Maybe we could rename this to `AppendLog` or `WriteLog`? The `fn` suffix doesn't add a lot of value.

File: upstream.go (hunk 23 above)
Comment (emersion) [reply upstream.go]: Ugh, GitHub shows these lines as added in both the NOTICE commit and the logging commit. Not sure why.

File: upstream.go [hunk 26]
```diff
@@ -603,6 +622,8 @@ func (uc *upstreamConn) handleMessage(msg *irc.Message) error {
 				delete(ch.Members, msg.Prefix.Name)
 				ch.Members[newNick] = membership
 			}
+
+			uc.Logfn(ch.Name, "*** %s is now known as %s", msg.Prefix.Name, newNick)
```
Comment (emersion) [inline upstream.go]: We should only log this if the user is part of the channel members.

**Comment by emersion** [comment]
Merged "Add downstream NOTICE support".

Apart from the comments, the meat of the "Add support for bouncer logs" patch looks good to me!

File: upstream.go (hunk 22 above)
Comment (delthas) [reply upstream.go]: in that case, `msg.Prefix.Name = verne.freenode.net`, `msg.Prefix.User = ""`, `msg.Prefix.Host = ""`. that's why i used `msg.Prefix.User == "" && msg.Prefix.Host == ""`

File: config/config.go (hunk 24 above)
Comment (delthas) [reply config/config.go]: Right, makes sense

File: upstream.go (hunk 25 above)
Comment (delthas) [reply upstream.go]: OK,  `AppendLog` is fine

File: upstream.go (hunk 26 above)
Comment (delthas) [reply upstream.go]: Right

File: upstream.go (hunk 23 above)
Comment (delthas) [reply upstream.go]: Should be fixed now.

File: upstream.go (hunk 22 above)
Comment (emersion) [reply upstream.go]: Ah, got it!

File: upstream.go [hunk 27]
```diff
@@ -1438,3 +1437,11 @@ func (uc *upstreamConn) AppendLog(entity string, format string, a ...interface{}
 		uc.logger.Printf("failed to log message to %q: %v", log.name, err)
 	}
 }
+
+func splitSpace(s string) []string {
```
Comment (emersion) [inline upstream.go]: Maybe we can just use `strings.Fields`?

File: upstream.go [hunk 28]
```diff
@@ -725,12 +756,19 @@ func (uc *upstreamConn) handleMessage(msg *irc.Message) error {
...
+		}
+
 		if msg.Prefix.Name == uc.nick {
 			uc.logger.Printf("quit")
 		}
 
 		for _, ch := range uc.channels {
 			delete(ch.Members, msg.Prefix.Name)
+
+			uc.AppendLog(ch.Name, "*** Quits: %s (%s@%s) (%s)", msg.Prefix.Name, msg.Prefix.User, msg.Prefix.Host, reason)
```
Comment (emersion) [inline upstream.go]: We need to check whether the user belongs to the channel, otherwise we'll append this to all channel logs.

File: downstream.go [hunk 29]
```diff
@@ -1077,6 +1077,46 @@ func (dc *downstreamConn) handleMessageRegistered(msg *irc.Message) error {
...
+		var upstreamChannels map[int64][]string
+		if len(msg.Params) > 0 {
+			upstreamChannels = make(map[int64][]string)
+			channels := strings.Split(msg.Params[0], ",")
+			for _, channel := range channels {
+				uc, upstreamChannel, err := dc.unmarshalEntity(channel)
+				if err != nil {
+					return err
+				}
+				if c, ok := upstreamChannels[uc.network.ID]; ok {
```
Comment (emersion) [inline downstream.go]: Can be simplified to:

```go
upstreamChannels[uc.network.ID] = append(upstreamChannels[uc.network.ID], upstreamChannel)
```

File: upstream.go [hunk 30]
```diff
@@ -155,6 +162,84 @@ func (uc *upstreamConn) isChannel(entity string) bool {
...
+					Prefix:  dc.srv.prefix(),
+					Command: irc.RPL_LISTEND,
+					Params:  []string{dc.nick, "End of /LIST"},
+				})
+			})
+		}
+		found = true
+		if !all {
+			delete(uc.pendingListDownstreamSet, pl.downstreamID)
+			uc.user.forEachUpstream(func(uc *upstreamConn) {
```
Comment (emersion) [inline upstream.go]: Since this function can only "unblock" a LIST command for the current upstream connection `uc`, do we really need to iterate over all upstream connections here?

Comment (delthas) [reply upstream.go]: Suppose we're connected to downstreams *a*, *b*; and the client does LIST twice.

Pending now schematically looks like this: `[[a, b], [a, b]]`. Now imagine *a* answers much faster than *b*. After *a* answers, we get: `[[b], [a, b]]`. We can't yet send the other LIST to *a* because it would be sent before the first RPL_LISTEND. When *b* finally replies, we get: `[[], [a, b]]`, ie `[[a, b]]`. We now need to resend both  `a` and `b` for the next LIST.

Hence the need to iterate over all upstreams.

File: user.go [hunk 31]
```diff
@@ -96,6 +96,15 @@ type user struct {
...
+	pendingLists     []pendingList
+}
+
+type pendingList struct {
+	downstreamID uint64
+	// list of pending per-upstream LIST requests and their corresponding LIST message parameters
+	pendingRequests map[int64][]string
```
Comment (emersion) [inline user.go]: Can we replace `[]string` with the whole `*irc.Message`? It would make it clearer what the items are.

Might as well rename `pendingRequests` to `pendingCommands`?

File: user.go [hunk 32]
```diff
@@ -96,6 +96,15 @@ type user struct {
...
 	networks        []*network
 	downstreamConns []*downstreamConn
+
+	pendingListsLock sync.Mutex
+	pendingLists     []pendingList
+}
+
+type pendingList struct {
+	downstreamID uint64
+	// list of pending per-upstream LIST requests and their corresponding LIST message parameters
```
Comment (emersion) [inline user.go]:
```go
// list of per-upstream LIST commands not yet sent
```

File: user.go [hunk 33]
```diff
@@ -96,6 +96,15 @@ type user struct {
 
 	networks        []*network
 	downstreamConns []*downstreamConn
+
+	pendingListsLock sync.Mutex
+	pendingLists     []pendingList
```
Comment (emersion) [inline user.go]:
```go
// LIST commands in progress
```

File: user.go (hunk 31 above)
Comment (delthas) [reply user.go]: Right, I found it confusing as well, at first I put a typedef for that but it was kinda overkill. `*irc.Message` is a good idea.

File: user.go (hunk 33 above)
Comment (emersion) [reply user.go]: Maybe we could even rename this to `pendingLISTs` to make it clear it's about LIST, and not just a list?

File: user.go (hunk 32 above)
Comment (delthas) [reply user.go]: `not yet complete`*? we only clear it when we receive the last RPL_LISTEND (or error, or close)

File: downstream.go [hunk 34]
```diff
@@ -1077,6 +1077,46 @@ func (dc *downstreamConn) handleMessageRegistered(msg *irc.Message) error {
...
+					upstreamChannels[uc.network.ID] = []string{upstreamChannel}
+				}
+			}
+		}
+
+		dc.user.pendingLists = append(dc.user.pendingLists, pl)
+		dc.forEachUpstream(func(uc *upstreamConn) {
+			var params []string
+			if upstreamChannels == nil {
+				params = []string{}
```
Comment (emersion) [inline downstream.go]: Nit: no need to initialize `params`, `params = nil` would do (or just do nothing).

File: upstream.go [hunk 35]
```diff
@@ -59,6 +59,10 @@ type upstreamConn struct {
 
 	saslClient  sasl.Client
 	saslStarted bool
+
+	// set of pending LIST requests, per downstream
```
Comment (emersion) [inline upstream.go]:
```go
// set of LIST commands in progress, per downstream
```

to make it clear what "pending" means (sent but not finished, as opposed to not yet sent)

File: user.go (hunk 32 above)
Comment (emersion) [reply user.go]: Eh, right! +1
